src/Microsoft.AspNetCore.Authentication.Google/GoogleHandler.cs

```diff
@@ -71,29 +71,35 @@ protected override string BuildChallengeUrl(AuthenticationProperties properties,
             return authorizationEndpoint;
         }
 
-        private static void AddQueryString(
+        private void AddQueryString<T>(
             IDictionary<string, string> queryStrings,
             AuthenticationProperties properties,
             string name,
-            string defaultValue = null)
+            Func<T, string> formatter,
+            T defaultValue)
         {
-            string value;
-            if (!properties.Items.TryGetValue(name, out value))
+            string value = null;
+            var parameterValue = properties.GetParameter<T>(name);
+            if (parameterValue != null)
             {
-                value = defaultValue;
+                value = formatter(parameterValue);
             }
-            else
+            else if (!properties.Items.TryGetValue(name, out value))
```

Comment:
Whoops, totally forgot about that part!

Reply:
We should add a test to ensure we didn't break this, maybe a test that customizes google challenge via Items and verifies that it doesn't show up in state?

Reply:
Yeah, I’ll adjust the tests to verify that.

Reply:
Added tests to all the Google challenge tests to verify that this behavior is guaranteed.

There was one undefinied behavior before, which was when a query argument was both specified through the `Parameters` *and* the `Items` collections. I had already given the `Pararameters` more weight before, since it’s our preferred way of overwriting these things. But it wasn’t clear whether the key should then still be removed from `Items` or not.

I’ve opted to remove it now although technically, the item was not consumed when building the query. I was really on the fence and wanted to argue against removing it to avoid introducing a side effect where the presence of a parameter will cause an item to be removed from the `Items` collection. But on a second thought, if the parameter was *not* there, it would have been removed anyway (and used as a query argument). So it makes more sense to just remove it.

Reply:
I think the important bit is that it gets removed from Items, which one wins doesn't really matter to me.  Whether it first applies Items and then applies Parameters, or in the opposite order, either way seems fine.